crates/kubelet/src/node.rs

```diff
@@ -389,3 +383,66 @@ fn lease_spec_definition(node_name: &str) -> serde_json::Value {
         }
     )
 }
+
+/// Defines the labels that will be applied to this node
+///
+/// Default values and passed node-labels arguments are injected by config.
+fn node_labels_definition(arch: &str, config: &Config) -> HashMap<String, String> {
+    // Add mandatory static labels
+    let mut labels = HashMap::new();
+    labels.insert("beta.kubernetes.io/os".to_string(), "linux".to_string());
+    labels.insert("kubernetes.io/os".to_string(), "linux".to_string());
+    labels.insert("kubernetes.io/role".to_string(), "agent".to_string());
+    labels.insert("type".to_string(), "krustlet".to_string());
+    // add the mandatory labels that are dependent on injected values
+    labels.insert("beta.kubernetes.io/arch".to_string(), arch.to_string());
+    labels.insert("kubernetes.io/arch".to_string(), arch.to_string());
+    labels.insert(
+        "kubernetes.io/hostname".to_string(),
+        config.hostname.to_string(),
+    );
+
+    let k8s_namespace = "kubernetes.io";
+    // namespaces managed by this method - do not allow user injection
+    let managed_namespace_labels = vec![
+        "beta.kubernetes.io/arch",
+        "beta.kubernetes.io/os",
+        "kubernetes.io/arch",
+        "kubernetes.io/hostname",
+        "kubernetes.io/os",
+        "kubernetes.io/role",
+        "type",
+    ];
+    let allowed_k8s_namespace_labels = vec![
+        "beta.kubernetes.io/instance-type",
+        "failure-domain.beta.kubernetes.io/region",
+        "failure-domain.beta.kubernetes.io/zone",
+        "failure-domain.kubernetes.io/region",
+        "failure-domain.kubernetes.io/zone",
+        "kubernetes.io/instance-type",
+    ];
+
+    // Attempt to append node labels from passed arguments.
+    // First, check for managed namespace and log exclusion
+    // Next, check if label contains k8s namespace and ensure it's allowable
+    // Else, if not k8s namspace, insert
+    let user_labels = config.node_labels.clone();
+
+    for (key, value) in user_labels.iter() {
+        if managed_namespace_labels.contains(&&key[..]) {
+            warn!(
+                "User provided node label {} omitted. Namespace label managed by runtime.",
+                key
+            );
+        } else if key.contains(k8s_namespace)
+            && (key.starts_with("kubelet.kubernetes.io")
+                || key.starts_with("node.kubernetes.io")
+                || allowed_k8s_namespace_labels.contains(&&key[..]))
```

Comment:
This doesn't look correct to me. I think we want to warn about invalid keys that don't follow the spec. This is what I'm thinking:

```diff
diff --git a/crates/kubelet/src/node.rs b/crates/kubelet/src/node.rs
index 9c69bd9..ae652a7 100644
--- a/crates/kubelet/src/node.rs
+++ b/crates/kubelet/src/node.rs
@@ -6,7 +6,7 @@ use k8s_openapi::apimachinery::pkg::apis::meta::v1::Time;
 use kube::api::{Api, DeleteParams, PatchParams, PostParams};
 use kube::error::ErrorResponse;
 use kube::Error;
-use log::{debug, error, info};
+use log::{debug, error, info, warn};
 
 macro_rules! retry {
     ($action:expr, times: $num_times:expr, error: $on_err:expr) => {{
@@ -339,13 +339,48 @@ fn node_definition(config: &Config, arch: &str) -> serde_json::Value {
     });
 
     // extra labels from config
-    for (key, val) in &config.node_labels {
+    for (key, val) in config.node_labels.iter().filter(|(k, _)| is_invalid_node_label_key(k)) {
         json["metadata"]["labels"][key] = serde_json::json!(val);
     }
 
     json
 }
 
+fn is_invalid_node_label_key(key: &str) -> bool {
+    let managed_namespace_labels = vec![
+        "beta.kubernetes.io/arch",
+        "beta.kubernetes.io/os",
+        "kubernetes.io/arch",
+        "kubernetes.io/hostname",
+        "kubernetes.io/os",
+        "kubernetes.io/role",
+        "type",
+    ];
+    let is_invalid = managed_namespace_labels.contains(&key) || is_invalid_k8s_node_label_key(key);
+
+    if is_invalid {
+        warn!("invalid label {}", key);
+    }
+
+    is_invalid
+}
+
+fn is_invalid_k8s_node_label_key(key: &str) -> bool {
+    let allowed_k8s_namespace_labels = vec![
+        "beta.kubernetes.io/instance-type",
+        "failure-domain.beta.kubernetes.io/region",
+        "failure-domain.beta.kubernetes.io/zone",
+        "failure-domain.kubernetes.io/region",
+        "failure-domain.kubernetes.io/zone",
+        "kubernetes.io/instance-type",
+    ];
+
+    key.contains("kubernetes.io")
+        && !(key.starts_with("kubelet.kubernetes.io")
+            || key.starts_with("node.kubernetes.io")
+            || allowed_k8s_namespace_labels.contains(&key))
+}
+
 /// Define a new coordination.Lease object for Kubernetes
 ///
 /// The lease tells Kubernetes that we want to claim the node for a while

```

Reply:
This is interesting.  Filter instead of builder....